internal/api/get_credentials_v1alpha1.go

```diff
@@ -0,0 +1,63 @@
+package api
+
+import (
+	"context"
+
+	"connectrpc.com/connect"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	typesv1alpha1 "github.com/akuity/kargo/internal/api/types/v1alpha1"
+	svcv1alpha1 "github.com/akuity/kargo/pkg/api/service/v1alpha1"
+)
+
+func (s *server) GetCredentials(
+	ctx context.Context,
+	req *connect.Request[svcv1alpha1.GetCredentialsRequest],
+) (*connect.Response[svcv1alpha1.GetCredentialsResponse], error) {
+	project := req.Msg.GetProject()
+	if err := validateFieldNotEmpty("project", project); err != nil {
+		return nil, err
+	}
+
+	name := req.Msg.GetName()
+	if err := validateFieldNotEmpty("name", name); err != nil {
+		return nil, err
+	}
+
+	if err := s.validateProjectExists(ctx, project); err != nil {
+		return nil, err
+	}
+
+	secret := corev1.Secret{}
+	if err := s.client.Get(
+		ctx,
+		types.NamespacedName{
+			Namespace: project,
+			Name:      name,
+		},
+		&secret,
+	); err != nil {
+		return nil, errors.Wrapf(err, "get secret %s", name)
+	}
+
+	secret = redactPassword(secret)
+
+	return connect.NewResponse(&svcv1alpha1.GetCredentialsResponse{
+		Credentials: typesv1alpha1.ToSecretProto(&secret),
+	}), nil
+}
+
+func redactPassword(secret corev1.Secret) corev1.Secret {
+	secret.StringData = make(map[string]string, len(secret.Data))
+	for k, v := range secret.Data {
+		if k == "password" {
+			secret.StringData[k] = "*** REDACTED ***"
+		} else {
+			secret.StringData[k] = string(v)
```

Comment:
Assuming someone for some reason decides to mix their Kargo credentials together with something else, this would leak the _other_ (non Kargo) secrets. Given this, I am wondering if we shouldn't work with a whitelist instead.

Reply:
Nice catch. imho, I should be checking for the existence of the credentials label here and treating the credential as not found if a secret is found lacking that label.

wdyt?

Reply:
Oh... also note that users cannot retrieve secrets this way that they aren't already authorized to get.

Reply:
Fixed in https://github.com/akuity/kargo/pull/1566/commits/c8b95af925db036dc83823b6aa0df182db391a5a

Reply:
With this patch, I think it's still odd (even with authorization checks) that we handle the password field with a mask while allowing other values to be plain text. I would either expect it to not have any masks, or to mask any fields except the ones which we expect to contain non-sensible values (e.g. `repoURL`, `username`, etc.)

Reply:
We expect Secrets that are credentials to have a payload that's structured a certain way. At present `password` is the only field we _expect_ that is sensitive.

But I'm starting to see your point, because suppose I had the following:

```
apiVersion: v1
kind: Secret
type: Opaque
metadata:
  name: kargo-demo
  namespace: kargo-demo
  labels:
    kargo.akuity.io/cred-type: git
stringData:
  repoURL: https://github.com/example/repo.git
  username: krancour
  passwrd: '12345'
```

__Notice that I fat-fingered the key for the password field.__

And created it like so:

```
kargo apply -f creds.yaml
```

Then `kargo get credentials -p kargo-demo kargo-demo` will show me the "passwd" unredacted.

So your point behind white-listing fields _not_ to redact is more clear to me now and I agree. Sorry it took me two tries to see the value in that.

I suppose it is still not _entirely_ foolproof, as nothing stops someone from, for instance, accidentally putting the password in the `username` field. (Whoops!) But I'm not sure there's anything that can be done about mistakes of that nature.

Reply:
Fixed in f4a8ecf54ab1c9c5fd9d3becd6df49d40e22ebaf